### PR TITLE
[terra-table] Rows and Header disabled checkbox accessibility fix

### LINKED ISSUE #2967: [terra-table] Single selectable table header row screen reader accessibility issue
# Bug Report
## Description
When focus is on terra table(single selection) header row and voice over on, header row checkbox(invisible) is supposed to be dimmed(/unselectable) but screen reader is reading "single selection, unchecked, checkbox column 1 of 3. To select this click option+control+space". This error occurs even when isDisabled prop set to true for header.

## Steps to Reproduce
1. With screen reader on, focus on header row of single selectable table example https://engineering.cerner.com/terra-ui/components/terra-table/table/guides/single-select-table
2. Use the above example and set isDisabled prop to true for headerData,
       selectAllColumn: {
              checkLabel: 'Single Selection',
              isDisabled: true,
            }

## Expected Behavior
Screen reader should read following or some more descriptive statement for header row. 
> "single selection, unchecked, checkbox, dimmed column 1 of 3. "

## Possible Solution
Seems like isDisabled prop is only spread to attribute that are visible.

## Environment
* Component Name and Version: Table/ terra-table 4.7.0
* Browser Name and Version: safari
* Operating System and version (desktop or mobile): Mac

### PR DESCRIPTION
**Pull request by lokesh-0813**
### Summary

Rows and Header checkbox disabled property was suppressed from being fed to screenreaders when they are visually hidden.
Changes in this PR allows screenreaders to access the disabled aria properties.
Updated existing guides.

Closes #2967 

### Deployment Link

https://terra-core-deployed-pr-3024.herokuapp.com/components/terra-table/table/table
### Testing

Tested in Safari with VoiceOver and IE/Chrome with JAWS.
### Additional Details

Thank you for contributing to Terra.
@cerner/terra

### CODE REVIEW

File: packages/terra-table/src/subcomponents/_HeaderCheckMarkCell.jsx [hunk 1]
```diff
@@ -112,6 +112,15 @@ const HeaderCheckMarkCell = ({
       attrSpread.onMouseDown = wrappedEventCallback(onMouseDown, event => event.currentTarget.setAttribute('data-cell-show-focus', 'false'));
     }
 
+    if (isSelected) {
```
Comment (dkasper-was-taken) [inline packages/terra-table/src/subcomponents/_HeaderCheckMarkCell.jsx:115]: We should be able to reduce the amount of duplication here. aria-checked and aria-disabled may be able to be set outside of the selectable check.

Comment (lokesh-0813) [reply packages/terra-table/src/subcomponents/_HeaderCheckMarkCell.jsx]: updated here 8d49a0f83d94b3b750db025fb58b2929599252ca

**Comment by dkasper-was-taken** [comment]
Nice work! We should add some examples to the guide documentation. I believe the single and multi-select examples likely need to be updated to set the disabled flag. Not sure about other examples, so you might check.

**Comment by lokesh-0813** [comment]
> Nice work! We should add some examples to the guide documentation. I believe the single and multi-select examples likely need to be updated to set the disabled flag. Not sure about other examples, so you might check.

Updated here 90b3e8019e5d72b80e8a9d5f5b20ac831bd801ce

**Comment by dkasper-was-taken** [comment]
> > Nice work! We should add some examples to the guide documentation. I believe the single and multi-select examples likely need to be updated to set the disabled flag. Not sure about other examples, so you might check.
> 
> Updated here [90b3e80](https://github.com/cerner/terra-core/commit/90b3e8019e5d72b80e8a9d5f5b20ac831bd801ce)

My mistake for not clarifying what I was expecting for examples. I don't believe that the examples need a toggle for disabling the rows. I would expect the `row` page to have a disabled row examples, but the guides are more for the ideal examples. What was missing for single and multi-select is the disabling of the CheckMarkHeaderCell similar to issue originally logged.

**Comment by lokesh-0813** [comment]
> > > Nice work! We should add some examples to the guide documentation. I believe the single and multi-select examples likely need to be updated to set the disabled flag. Not sure about other examples, so you might check.
> > 
> > 
> > Updated here [90b3e80](https://github.com/cerner/terra-core/commit/90b3e8019e5d72b80e8a9d5f5b20ac831bd801ce)
> 
> My mistake for not clarifying what I was expecting for examples. I don't believe that the guides need a toggle for disabling the rows. I would expect the `row` page to have a disabled row examples, but the guides are more for the ideal examples. What was missing for single and multi-select is the disabling of the CheckMarkHeaderCell similar to issue originally logged.

Updated row examples here 47ce209a111eae005f2d3a5c5753b586988ae5d6

There is already a disabled header example in `header` page.

**Comment by jeremyfuksa** [comment]
This is working for me in VoiceOver, but I'm not getting it to work with JAWS. It just skips over the disabled items. Tips on how you got it to work in JAWS, @lokesh-0813 ?

**Comment by lokesh-0813** [comment]
> This is working for me in VoiceOver, but I'm not getting it to work with JAWS. It just skips over the disabled items. Tips on how you got it to work in JAWS, @lokesh-0813 ?

Hey Jeremy, sorry for the delayed response I am having trouble with the citrix box since couple of days. I had to take @supreethmr 's help test again and makes sure it's working fine.

Here is the workflow
-> Navigate to Guides -> Single Select Table/Multi Select Table
-> Select first row while JAWS On.
-> Use up arrow key to navigate to column names and navigate till you reach the invisible check mark(one arrow up after you have "Column 0" highlighted).

This is what I have tried and it has worked for me. Let me know if you need more evidence. I will try to attach a video once my citrix receiver issue is resolved.

File: packages/terra-table/src/terra-dev-site/doc/table/guides.5/MultiSelectTable.2.doc.mdx [hunk 2]
```diff
@@ -92,6 +92,8 @@ We can then implement the unpack of our data into our row cells.
 ```
 Finally we can implement a method to loop through our data and create the table row with our methods and call it from our render method.
 Ensuring that we add the appropriate aria label for multiple selections.
+Additionally whenever the table is rendered without a `check mark`, for instance when rowStyle="toggle" it is recommended to set isDisabled=true
+property provided by the `selectAllColumn` object.
```
Comment (yuderekyu) [inline packages/terra-table/src/terra-dev-site/doc/table/guides.5/MultiSelectTable.2.doc.mdx:96]:
```suggestion
```suggestion
Additionally, set `isDisabled=true` within the `selectAllColumn` property whenever the table is rendered without a checkmark. An example is when `rowStyle="toggle"`.
```
```

File: packages/terra-table/src/terra-dev-site/doc/table/guides.5/SingleSelectTable.1.doc.mdx [hunk 3]
```diff
@@ -77,6 +77,9 @@ We can then implement the unpack of our data into our row cells.
 + const createCellsForRow = cells => cells.map(cell => createCell(cell));
 ```
 Finally we can implement a method to loop through our data and create the table with our methods and call it from our render method.
+Additionally whenever the table is rendered without a `check mark`, for instance when rowStyle="toggle" it is recommended to set isDisabled=true
```
Comment (yuderekyu) [inline packages/terra-table/src/terra-dev-site/doc/table/guides.5/SingleSelectTable.1.doc.mdx:80]:
```suggestion
Additionally, set `isDisabled=true` within the `selectAllColumn` property whenever the table is rendered without a checkmark. An example is when `rowStyle="toggle"`.
```

**Comment by dkasper-was-taken** [comment]
@lokesh-0813 I believe what @jeremyfuksa was referring to is that when navigating to the table initially with jaws, either through use of the 't' key, or stepping into the table sequentially the column header for selection column is skipped, and only navigation backwards to it will read the column. From some experimentation it looks like correcting the hidden css corrects this:

CheckMarkCell.module.scss
HeaderCheckMarkCell.module.scss
```
  .checkmark.is-hidden {
    background-color: transparent;
    background-image: none;
    border: 0;
    height: 1; // this value
    margin: -1; // this value
    padding: 0;
    width: 1; // this value
  }
```

**Comment by lokesh-0813** [comment]
> @lokesh-0813 I believe what @jeremyfuksa was referring to is that when navigating to the table initially with jaws, either through use of the 't' key, or stepping into the table sequentially the column header for selection column is skipped, and only navigation backwards to it will read the column. From some experimentation it looks like correcting the hidden css corrects this:
> 
> CheckMarkCell.module.scss
> HeaderCheckMarkCell.module.scss
> 
> ```
>   .checkmark.is-hidden {
>     background-color: transparent;
>     background-image: none;
>     border: 0;
>     height: 1; // this value
>     margin: -1; // this value
>     padding: 0;
>     width: 1; // this value
>   }
> ```

@dkasper-was-taken updated here 721d39e47391a2b6401c900fe303bdbe0803e0e0
@jeremyfuksa  Let me know if you still face problems testing in JAWS.

**Comment by lokesh-0813** [comment]
> +1 assuming the updated screenshot is still valid.

There is a minor change to the rows with addition of `width: 1`. I was thinking of making it display:inline or display:none.

**Comment by dkasper-was-taken** [comment]
@lokesh-0813 Trying setting the position to 'absolution' first. Depending on how that calculates it may top need a left/top set. Setting "display: none" would cause it no to be read by the screen reader.

**Comment by lokesh-0813** [comment]
> @lokesh-0813 Trying setting the position to 'absolute' first. Depending on how that calculates it may top need a left/top set. Setting "display: none" would cause it no to be read by the screen reader.

Does the job. Updated here 01b32ba86ba0feeb52f1b54f9929740160d2b03b

File: packages/terra-table/CHANGELOG.md [hunk 4]
```diff
@@ -12,6 +12,7 @@ Unreleased
 4.8.0 - (June 9, 2020)
 ------------------
 ### Changed
+* Rows and Header checkbox when disabled are now accessible for screenreaders to read out accordingly.
```
Comment (avinashg1994) [inline packages/terra-table/CHANGELOG.md:15]: Move this log to the "Unreleased" section

Comment (lokesh-0813) [reply packages/terra-table/CHANGELOG.md]: updated

**Comment by ryanthemanuel** [comment]
@dkasper-was-taken is checking with @scottwilmarth about a few things before we merge this.  Marking it with do not merge until that gets resolved.

**Comment by dkasper-was-taken** [comment]
@lokesh-0813 After checking with @scottwilmarth, when header column is not intended for interaction the checkbox should not be present and the only thing that should be read is the title of the cell. A disabled state is still possible, but that should leverage both selectable & disabled when face up when the whole table is displayed as disabled.

I created an alternate branch from this one with the adjusted changes:
https://github.com/cerner/terra-core/tree/table-screenreader-rows-revised

**Comment by ryanthemanuel** [comment]
As an FYI, here's a link with the diffs for @dkasper-was-taken's revised branch: https://github.com/cerner/terra-core/compare/table-screenreader-rows...table-screenreader-rows-revised Does this seem reasonable @lokesh-0813?

**Comment by lokesh-0813** [comment]
@dkasper-was-taken I have pulled the [revised](https://github.com/cerner/terra-core/tree/table-screenreader-rows-revised) branch. I have gone through the changes and did some testing too. It looks fine to me. 
@ryanthemanuel  Updated here 2897832c0e81bfbb2837735c9a072fe2edefa46b

**Pull request merged**